File: web/src/contract.gen.ts
// =============================================================================
// GENERIERT — NICHT VON HAND ÄNDERN.
// Quelle: docs/contract/api-contract.json (aus der laufenden API aufgezeichnet).
// Neu erzeugen: node verifikation/contract_gen_ts.mjs
// Frischeprüfung: scripts/verify-contract.sh (Gate im QA-Loop)
//
// Zweck: Die PWA leitet ihre Typen HIER ab statt sie zweitzuschreiben. Benennt
// der Server ein Feld um, ändert sich diese Datei — und jede Lesestelle im
// Client wird zum Compile-Fehler. Das ist die strukturelle Antwort auf die drei
// Launch-Blocker vom 21.08.2026, die alle Vertragsdrift waren.
// =============================================================================
/* eslint-disable */

/** POST /v1/billing/activate */
export type BillingActivateResponse = {
  access_token: string;
  restore_code: string;
  tier: string;
  valid_until: string;
};

/** POST /v1/billing/activate (Beleg erneut) */
export type BillingActivateKonfliktResponse = {
  error: {
    code: string;
    message: string;
  };
};

/** POST /v1/billing/restore
  * Restore ROTIERT den Zugangstoken — der Token des alten Geraets wird dabei ungueltig. Beabsichtigt (ein Abo, ein Geraet), aber ein Client darf nach einem Restore nicht mehr mit dem alten Token weiterarbeiten. */
export type BillingRestoreResponse = {
  access_token: string;
  tier: string;
};

/** POST /v1/billing/validate */
export type BillingValidateResponse = {
  status: string;
  tier: string;
  valid_until: string;
};

/** POST /v1/billing/validate (alter Token nach Restore) */
export type BillingValidateAbgelaufenResponse = {
  error: {
    code: string;
    message: string;
  };
};

/** GET /v1/cities/{slug}/alerts */
export type CitiesAlertsResponse = unknown[];

/** GET /v1/cities/{slug}/events */
export type CitiesEventsResponse = Array<{
  created_at: string;
  event_id: number;
  report_id: string;
  type: string;
}>;

/** GET /v1/cities */
export type CitiesListResponse = Array<{
  display_name: string;
  is_active: boolean;
  slug: string;
}>;

/** POST /v1/devices */
export type DevicesIssueResponse = {
  device_id: string;
  device_token: string;
};

/** GET /v1/devices/me */
export type DevicesMeResponse = {
  access: string;
  rank: string;
  subscription_price_eur: number;
  total_reports: number;
  trial_days_remaining: number;
  trust_score: number;
};

/** GET /v1/cities/{slug}/stops/nearby (Fehlerfall) */
export type FehlerValidationResponse = {
  error: {
    code: string;
    message: string;
  };
};

/** GET /health/live */
export type HealthLiveResponse = {
  status: string;
};

/** GET /health/ready */
export type HealthReadyResponse = {
  status: string;
};

/** POST /v1/journeys/search
  * next_departures traegt KEIN route_id/headsign — die stehen auf der Verbindung. Genau diese Asymmetrie war Launch-Blocker 2 (Flutter-TypeError). */
export type JourneysSearchResponse = {
  direct_connections: Array<{
    direction_id: number;
    headsign: string;
    next_departures: Array<{
      realtime: boolean;
      scheduled_time: string;
      trip_ref: {
        start_date: string;
        trip_id: string;
      };
      warnings: unknown[];
    }>;
    ride_seconds: number;
    route_id: string;
    stops_count: number;
  }>;
  from_stop_id: string;
  to_stop_id: string;
  transfer_connections: Array<{
    leg_a: {
      alight_at: string;
      alight_stop: string;
      board_at: string;
      board_stop: string;
      headsign: string;
      route_id: string;
    };
    leg_b: {
      alight_at: string;
      alight_stop: string;
      board_at: string;
      board_stop: string;
      headsign: string;
      route_id: string;
    };
    total_seconds: number;
    transfer_stop_id: string;
    wait_seconds: number;
  }>;
};

/** GET /v1/journeys/{tripId}/warnings */
export type JourneysWarningsResponse = unknown[];

/** POST /v1/reports */
export type ReportsCreateResponse = {
  anchor_type: string;
  city_slug: string;
  created_at: string;
  expires_at: string;
  id: string;
  origin: string;
  report_type: string;
  station_name: string;
  status: string;
};

/** POST /v1/reports/{id}/events */
export type ReportsEventResponse = null;

/** GET /v1/cities/{slug}/reports
  * route_id/headsign/trip_id nur bei Trip-Anker — bei Stations-Anker fehlen sie (WhenWritingNull). Deshalb optional. */
export type ReportsListResponse = Array<{
  anchor_type: string;
  city_slug: string;
  created_at: string;
  expires_at: string;
  headsign?: string;
  id: string;
  origin: string;
  report_type: string;
  route_id?: string;
  station_name: string;
  status: string;
  trip_id?: string;
  trip_start_date?: string;
}>;

/** GET /v1/stops/{stopId}/departures */
export type StopsDeparturesResponse = Array<{
  headsign: string;
  realtime: boolean;
  route_id: string;
  scheduled_time: string;
  trip_ref: {
    start_date: string;
    trip_id: string;
  };
  warnings: unknown[];
}>;

/** GET /v1/cities/{slug}/stops/nearby
  * distance_km nur bei lat/lon-Suche — Namenssuche liefert es nicht. */
export type StopsNearbyResponse = Array<{
  distance_km?: number;
  lat: number;
  lon: number;
  stop_id: string;
  stop_name: string;
}>;

/** Pflicht-Header je Endpunkt — im Vertrag durch Negativproben belegt. */
export const PFLICHT_HEADER = {
  "billing.restore": [
    "X-Device-Token"
  ],
  "cities.events": [
    "X-Device-Token",
    "X-Ticket-Confirmed"
  ],
  "devices.me": [
    "X-Device-Token"
  ],
  "journeys.warnings": [
    "X-Device-Token",
    "X-Ticket-Confirmed"
  ],
  "reports.create": [
    "X-Device-Token",
    "Idempotency-Key"
  ],
  "reports.list": [
    "X-Device-Token",
    "X-Ticket-Confirmed"
  ]
} as const;

File: app/lib/contract.gen.dart
// =============================================================================
// GENERIERT — NICHT VON HAND ÄNDERN.
// Quelle: docs/contract/api-contract.json (aus der laufenden API aufgezeichnet).
// Neu erzeugen: node verifikation/contract_gen_dart.mjs
// Frischeprüfung: scripts/verify-contract.sh (Gate im QA-Loop)
// =============================================================================

/// Header, ohne die der Server ablehnt — im Vertrag durch Negativproben belegt.
const kVertragPflichtHeader = <String, List<String>>{
  'billing.restore': ['X-Device-Token'],
  'cities.events': ['X-Device-Token', 'X-Ticket-Confirmed'],
  'devices.me': ['X-Device-Token'],
  'journeys.warnings': ['X-Device-Token', 'X-Ticket-Confirmed'],
  'reports.create': ['X-Device-Token', 'Idempotency-Key'],
  'reports.list': ['X-Device-Token', 'X-Ticket-Confirmed'],
};

/// Feldpfade, die der Server IMMER sendet (in jeder Beobachtung vorhanden).
const kVertragPflichtFelder = <String, List<String>>{
  'billing.activate': ['access_token', 'restore_code', 'tier', 'valid_until'],
  'billing.activate.konflikt': ['error', 'error.code', 'error.message'],
  'billing.restore': ['access_token', 'tier'],
  'billing.validate': ['status', 'tier', 'valid_until'],
  'billing.validate.abgelaufen': ['error', 'error.code', 'error.message'],
  'cities.alerts': ['[]', '[].description', '[].header', '[].route_refs', '[].severity'],
  'cities.events': ['[]', '[].created_at', '[].event_id', '[].report_id', '[].type'],
  'cities.list': ['[]', '[].display_name', '[].is_active', '[].slug'],
  'devices.issue': ['device_id', 'device_token'],
  'devices.me': ['access', 'rank', 'subscription_price_eur', 'total_reports', 'trial_days_remaining', 'trust_score'],
  'fehler.validation': ['error', 'error.code', 'error.message'],
  'health.live': ['status'],
  'health.ready': ['status'],
  'journeys.search': ['direct_connections', 'direct_connections[]', 'direct_connections[].direction_id', 'direct_connections[].headsign', 'direct_connections[].next_departures', 'direct_connections[].next_departures[].realtime', 'direct_connections[].next_departures[].scheduled_time', 'direct_connections[].next_departures[].trip_ref', 'direct_connections[].next_departures[].trip_ref.start_date', 'direct_connections[].next_departures[].trip_ref.trip_id', 'direct_connections[].next_departures[].warnings', 'direct_connections[].next_departures[].warnings[].affected_stop_id', 'direct_connections[].next_departures[].warnings[].affected_stop_sequence', 'direct_connections[].next_departures[].warnings[].message', 'direct_connections[].next_departures[].warnings[].user_eta_seconds', 'direct_connections[].ride_seconds', 'direct_connections[].route_id', 'direct_connections[].stops_count', 'from_stop_id', 'to_stop_id', 'transfer_connections', 'transfer_connections[]', 'transfer_connections[].leg_a', 'transfer_connections[].leg_a.alight_at', 'transfer_connections[].leg_a.alight_stop', 'transfer_connections[].leg_a.board_at', 'transfer_connections[].leg_a.board_stop', 'transfer_connections[].leg_a.headsign', 'transfer_connections[].leg_a.route_id', 'transfer_connections[].leg_b', 'transfer_connections[].leg_b.alight_at', 'transfer_connections[].leg_b.alight_stop', 'transfer_connections[].leg_b.board_at', 'transfer_connections[].leg_b.board_stop', 'transfer_connections[].leg_b.headsign', 'transfer_connections[].leg_b.route_id', 'transfer_connections[].total_seconds', 'transfer_connections[].transfer_stop_id', 'transfer_connections[].wait_seconds'],
  'journeys.warnings': ['[]', '[].affected_stop_id', '[].affected_stop_sequence', '[].message', '[].report_id', '[].user_eta_seconds'],
  'reports.create': ['anchor_type', 'city_slug', 'created_at', 'expires_at', 'id', 'origin', 'report_type', 'station_name', 'status'],
  'reports.event': [],
  'reports.list': ['[]', '[].anchor_type', '[].city_slug', '[].created_at', '[].expires_at', '[].id', '[].origin', '[].report_type', '[].station_name', '[].status'],
  'stops.departures': ['[]', '[].headsign', '[].realtime', '[].route_id', '[].scheduled_time', '[].trip_ref', '[].trip_ref.start_date', '[].trip_ref.trip_id', '[].warnings'],
  'stops.nearby': ['[]', '[].lat', '[].lon', '[].stop_id', '[].stop_name'],
};

/// Feldpfade, die FEHLEN koennen (WhenWritingNull oder anker-/abfrageabhaengig).
const kVertragOptionaleFelder = <String, List<String>>{
  'billing.activate': [],
  'billing.activate.konflikt': [],
  'billing.restore': [],
  'billing.validate': [],
  'billing.validate.abgelaufen': [],
  'cities.alerts': ['[].route_refs[]', '[].url'],
  'cities.events': [],
  'cities.list': [],
  'devices.issue': [],
  'devices.me': [],
  'fehler.validation': [],
  'health.live': [],
  'health.ready': [],
  'journeys.search': ['direct_connections[].next_departures[]', 'direct_connections[].next_departures[].warnings[]'],
  'journeys.warnings': [],
  'reports.create': [],
  'reports.event': [],
  'reports.list': ['[].headsign', '[].route_id', '[].trip_id', '[].trip_start_date'],
  'stops.departures': [],
  'stops.nearby': ['[].distance_km'],
};

/// Endpunkte, deren Antwortform die Aufzeichnung NICHT sehen konnte
/// (leere Liste bzw. null). Bewusst ausgewiesen statt stillschweigend als
/// „keine Felder" verbucht — hier schuetzt der Vertrag nichts.
const kVertragUnbeobachtet = <String>[];
